File: frontend/e2e/helpers/agentOrigin.js
"use strict";

/**
 * Notice 39 / #4763 — the e2e rail names itself on our own wire.
 *
 * ## The rail this exists for, and what it was doing
 *
 * `backend/app/utils/agent_origin.py` is this contract for the ~90 Python probes.
 * It cannot reach here: it is Python, and `browser-audit.yml` drives Node. So the
 * one fleet rail that runs against production ON A SCHEDULE was the one rail with
 * no carrier at all.
 *
 * MEASURED on production, 30-day window, read 2026-09-10 (#4763). The
 * `tournament-inventory` pack resolves two slugs by asking the real search API
 * (`/api/events/search?q=tennis`, `?q=grand%20prix`), on the `desktop` and
 * `mobile` projects, once per scheduled run. Only CI asks for BOTH terms, so
 * co-occurrence inside an hour separates robot from person:
 *
 *     grand prix   51 rows in a CI hour,  0 solo   — rank 17 of 571 terms
 *     tennis       51 rows in a CI hour, 12 solo   — rank 11 of 571 terms
 *
 * `grand prix` has ZERO organic searches in thirty days. It sits in the top 20 of
 * the table notice 39 names — "every 'what are people doing' read must tell a
 * person from a lane" — purely because we ask for it once a day.
 *
 * ## What this does NOT claim, because it was checked
 *
 * No warming or latency harm today. `search_head_warmer._head_from_user_rows`
 * filters to `session_id IS NOT NULL OR user_id IS NOT NULL`, and both terms
 * measure 100% session-less (0 attested of 51 and of 63), so the head warmer
 * cannot elect either one; `DEFAULT_HEAD_SIZE = 8` puts them below the cut in any
 * case. `/api/events/search` casts no trending vote either — both
 * `_record_trending` call sites are inside `typeahead_search`.
 *
 * So the defect is that the TABLE misreports demand to anything reading it without
 * knowing to filter on `session_id`. The designed defense is this header; what is
 * actually covering us is an unrelated filter that happens to exclude session-less
 * rows. That is worth closing before a "top search terms" panel inherits the lie.
 *
 * ## Why a helper and not `extraHTTPHeaders` in playwright.config.ts
 *
 * `tools/shop-shot.mjs` sets the header that way and is right to: it shoots our own
 * pages and nothing else. This rail is different in two ways that both bite.
 *
 *   1. A context-wide header goes to EVERY host a page contacts, including the
 *      third parties our pages load. An internal header naming our lanes has no
 *      business on a third party's wire — the same rule `is_our_host` exists to
 *      enforce in Python, and the same rule `agentOriginCarriers.contract.test.js`
 *      already pins for the shell carrier ("never sends the header to a host that
 *      is not ours").
 *   2. The alternative — a `page.route()` interceptor that filters by host — puts
 *      a JS hop in front of every request in a rail that contains
 *      `discover-latency.spec.ts` and `sports-latency.spec.ts`. Distorting the
 *      thing you are measuring to label it is not a trade worth making.
 *
 * So the tag goes on the EXPLICIT API calls the specs make, all of which are built
 * from `AUDIT_API_BASE_URL` and are therefore ours by construction — and
 * `isOurHost` still guards them, so the guarantee does not rest on that being
 * remembered.
 *
 * ## The naming rule differs from Python's ON PURPOSE
 *
 * `resolve_agent()` in Python returns None when `BL_AGENT` is unset, and MUST: an
 * unset variable is an ABSENCE, a human running a script by hand looks exactly
 * like a lane that forgot, and tagging on absence would silently delete real
 * people from the table (notice 39 guard 1).
 *
 * `GITHUB_ACTIONS === "true"` is not an absence. It is a positive assertion, made
 * by the runner and not by us, that no person is present. A run inside GitHub
 * Actions is definitionally a robot, so naming it is a deduction rather than a
 * guess — and it means a workflow added next month is tagged without anyone
 * remembering to edit an `env:` block. `BL_AGENT` still wins when set, so a human
 * driving a pack locally is untagged, and `BL_AGENT=user` still says "count me".
 */

/** The wire name. Must agree with `agent_origin.ORIGIN_HEADER` character for character. */
const ORIGIN_HEADER = "x-bainluck-origin";

/** The one value honoured POSITIVELY — it KEEPS the search-log row. */
const ORIGIN_USER = "user";

const OUR_HOSTS = ["bainluck.com", "localhost", "127.0.0.1"];

/**
 * The agent name for THIS call, or null when nothing may be added.
 *
 * Read at call time, never captured at module load: `agent_origin.resolve_agent`
 * carries a long comment about why, and it applies identically here. A name baked
 * in at import produces the wrong answer for any caller that sets `BL_AGENT`
 * afterwards, and the failure is silent in both directions.
 */
function resolveAgent() {
  const named = (process.env.BL_AGENT || "").trim();
  if (named) return named;

  // Not a default-on-absence. See the header comment: this is the runner
  // asserting that it is a runner.
  if (process.env.GITHUB_ACTIONS === "true") {
    const workflow = (process.env.GITHUB_WORKFLOW || "").trim();
    // Kept to the shape a log reader can scan: `ci:Browser audit` → `ci:browser-audit`.
    const slug = workflow
      ? workflow.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-+|-+$/g, "")
      : "";
    return slug ? `ci:${slug}` : "ci";
  }

  return null;
}

/** Lowercased hostname of `url`, or "" when it has none. */
function hostOf(url) {
  try {
    // A bare host has no protocol until it is given one, rather than being
    // string-munged — the same reason `agent_origin._host_of` does this.
    return new URL(String(url).includes("//") ? url : `//${url}`, "http://placeholder").hostname.toLowerCase();
  } catch {
    // An unparseable URL is not ours; refusing to tag is the safe direction.
    return "";
  }
}

/**
 * True when `url` points at a host we own.
 *
 * Matches the parsed HOST as a suffix, never a substring: a substring test tags
 * `https://example.com/?ref=bainluck.com`, which is a third party.
 */
function isOurHost(url) {
  const host = hostOf(url);
  if (!host) return false;
  return OUR_HOSTS.some((h) => host === h || host.endsWith(`.${h}`));
}

/**
 * Headers to ADD so `url` says who sent it. Possibly empty — the caller merges.
 *
 * Empty is the correct answer for a third-party host, an already-tagged request,
 * or a caller that never named itself. An explicit header from the caller always
 * wins: two disagreeing `x-bainluck-origin` values are worse than none, because
 * the backend reads the first and the caller's stated intent loses silently.
 */
function originHeaders(url, existing) {
  if (!isOurHost(url)) return {};

  const lowered = new Set(Object.keys(existing || {}).map((k) => String(k).toLowerCase()));
  if (lowered.has(ORIGIN_HEADER)) return {};

  const who = resolveAgent();
  if (who === null) return {};

  const headers = { [ORIGIN_HEADER]: who };
  // The router log records the UA, which makes fleet traffic legible without a
  // database read. Never clobber a caller's own — and note this rail CAN set one
  // safely, unlike `shop-shot.mjs`, because these are API calls and not the page
  // render whose fidelity a bot UA would change.
  if (!lowered.has("user-agent")) {
    headers["User-Agent"] = `BainLuckBot/1.0 (${who})`;
  }
  return headers;
}

/** `headers` plus the origin tag — the one-call form for request builders. */
function tagged(url, headers) {
  const merged = { ...(headers || {}) };
  return { ...merged, ...originHeaders(url, merged) };
}

module.exports = {
  ORIGIN_HEADER,
  ORIGIN_USER,
  resolveAgent,
  isOurHost,
  originHeaders,
  tagged,
};

File: backend/app/utils/agent_origin.py
"""Outbound `x-bainluck-origin` tagging for our own Python probes.

Notice 39 / #1916: every production read our fleet makes should say whether a
person or a lane made it. `tools/bl-agent-curl.sh` (rung 1) does that for the
`curl` an agent types by hand. This module is the same contract for the other
half of the fleet's traffic — the Python scripts under `backend/scripts/` that
call `/api/admin/db-query` — which no shell shadow can reach.

WHAT THE TAG BUYS HERE, STATED HONESTLY
---------------------------------------
On the SEARCH route the tag has teeth: `routes/events.py:_request_is_automation`
suppresses the search-query log write and the trending vote, so an untagged
probe does not merely add a row, it VOTES, and the head warmer then spends real
work warming our own specimens.

The gate scripts call `/api/admin/db-query`, so none of that applies to them.
Their tag is ATTRIBUTION ONLY, and deliberately so:

  * It buys them no rate-limit relief. Admin paths are keyed on a hash of the
    admin token (300/min) and that bucket outranks the trusted-address ceiling
    in `utils/rate_limit.py`, so tagging cannot and must not change what a gate
    is allowed to spend.
  * The header is caller-supplied and therefore forgeable. It is fit for saying
    WHO, never for granting WHAT — see the "CEILING, NEVER EXEMPTION" note in
    `utils/rate_limit.py`. Nothing in this module should ever be read by an
    authorization path.

THE HEADER NAME LIVES HERE, ONCE
--------------------------------
`routes/events.py` reads this constant rather than its own literal. A sender and
a reader that disagree by one character produce no error anywhere: every probe
looks tagged, every row still gets written, and the drift is invisible until
someone counts. One definition removes that failure mode by construction.
"""

from __future__ import annotations

import os
from typing import Dict, Mapping, Optional
from urllib.parse import urlsplit

#: The wire name. `routes/events.py` imports this; do not re-spell it anywhere.
ORIGIN_HEADER = "x-bainluck-origin"

#: The one value that is honoured POSITIVELY — it keeps the search-log row.
#: Use it when you mean to measure as a person.
ORIGIN_USER = "user"

_OUR_HOSTS = ("bainluck.com", "localhost", "127.0.0.1")


def resolve_agent() -> Optional[str]:
    """The agent name for THIS call, read from ``BL_AGENT`` at call time, or None.

    Read at call time, never captured at import time. The shell helper shipped
    with a comment about this because the obvious shape fails silently: a name
    baked in at load produces an EMPTY header, the backend's ``if not raw``
    reads empty as a person, and the call votes in the search head exactly as it
    did untagged — while the author has every reason to believe it is tagged.
    A module-level constant here would reintroduce that for any script that sets
    ``BL_AGENT`` after import.

    UNSET OR BLANK RETURNS None, AND THE CALLER THEN ADDS NOTHING (notice 39
    guard 1). An earlier draft substituted ``agent-unnamed`` on the reasoning
    that being unnamed is not evidence of humanity. That reasoning is right
    about the WORLD and wrong about the DIRECTION of the failure, which is what
    decides a default:

      * Tagging an unnamed caller SUPPRESSES its search-log row
        (`routes/events.py:_request_is_automation` treats any non-"user" value
        as automation). So a default that tags makes every un-configured caller
        — a human running a script by hand included — vanish from the table.
        That drains the warm head silently and leaves nothing behind to notice.
      * Not tagging leaves a row we can see, count and later attribute.

    `_request_is_automation`'s own docstring picks the same direction for the
    same reason: it FAILS TOWARD LOGGING. A default is a guess, and the only
    safe guess is the one whose mistakes are visible.
    """
    return (os.environ.get("BL_AGENT") or "").strip() or None


def _host_of(url: str) -> str:
    """Lowercased hostname of ``url``, or "" if it has none."""
    try:
        # A scheme-relative or bare-host string has no netloc until it has a
        # scheme, so give it one rather than string-munging the URL by hand.
        split = urlsplit(url if "//" in url else f"//{url}")
        return (split.hostname or "").lower()
    except ValueError:
        # An unparseable URL is not ours; refusing to tag is the safe direction.
        return ""


def is_our_host(url: str) -> bool:
    """True when ``url`` points at a host we own.

    Our probes call Kalshi, Polymarket, ESPN and GitHub from the same scripts.
    An internal header naming our lanes has no business on a third party's wire,
    so this matches the parsed HOST as a suffix. A substring test would have
    tagged ``https://example.com/?ref=bainluck.com``.
    """
    host = _host_of(url)
    if not host:
        return False
    return any(host == h or host.endswith(f".{h}") for h in _OUR_HOSTS)


def origin_headers(
    url: str, existing: Optional[Mapping[str, str]] = None
) -> Dict[str, str]:
    """Headers to ADD so ``url`` says who sent it. Possibly empty.

    Returns a new dict; the caller merges it. Empty means "add nothing", which
    is the correct answer for a third-party host, an already-tagged request, or
    a caller who never named itself (see `resolve_agent`).

    An explicit header from the caller always wins. A request carrying two
    `x-bainluck-origin` values that disagree about whether its sender is a
    person is worse than one carrying none: the backend reads ``.get()``, i.e.
    the first, so the caller's stated intent could silently lose to this
    module's default.
    """
    if not is_our_host(url):
        return {}

    lowered = {str(k).lower() for k in (existing or {})}
    if ORIGIN_HEADER in lowered:
        return {}

    who = resolve_agent()
    if who is None:
        # Pass through untouched: no origin header AND no bot User-Agent. Both
        # halves matter — a `BainLuckBot/1.0` UA with no origin header is a
        # request that looks automated in the router log while still voting in
        # the search head, which is the worst of both readings.
        return {}

    headers = {ORIGIN_HEADER: who}
    # The router log records the UA, which is what makes fleet traffic legible
    # in a log window without a database read. Never clobber a caller's own.
    if "user-agent" not in lowered:
        headers["User-Agent"] = f"BainLuckBot/1.0 ({who})"
    return headers


def tagged(url: str, headers: Optional[Mapping[str, str]] = None) -> Dict[str, str]:
    """``headers`` plus the origin tag — the one-call form for request builders."""
    merged = dict(headers or {})
    merged.update(origin_headers(url, merged))
    return merged
